### ISSUE TITLE
Next release versioning proposal

### ISSUE THREAD

**Issue by henryiii** [pypa/cibuildwheel#2047]
### Description

I'd like to make a proposal for the versioning strategy and contents of the next few cibuildwheel versions. Here's the suggestion:

The next release (2.22) has the following features:

* #2045 - But the default set of versions is the same if you don't set `enable`. We can warn if prerelease or free-threading standalone options are set.
* Warning for Python < 3.11 as host.

The following release of cibuildwheel I propose we call "3.0", with the following features:

* More `enable` groups, and removing the prerelease and free-threading standalone options.
* #1912

After this point, we can shift our numbering up one, with major releases for adding/removing Pythons, minor release for features, and patch releases for bug fixes. This would allow the much-requested `vX` GHA tag to be something we can provide.

Thoughts?

**Comment by joerick** [comment]
> The following release of cibuildwheel I propose we call "3.0", with the following features (not a complete list):
> 
> * More `enable` groups, and removing the prerelease and free-threading standalone options.
> * [feat: use SPEC 0 schedule for cibuildwheel #1912](https://github.com/pypa/cibuildwheel/pull/1912)

This sounds good. It especially works well with the introduction of the `enable` option - we can make some braver decisions there, such as dropping pypy from defaults, or changing the policy on building non-eol pythons by default if we like.

> * New version of [Split file utils into dedicated submodule #1873](https://github.com/pypa/cibuildwheel/pull/1873)

Ah yes. I'm happy to take a pass at this if you haven't started already?

> After this point, we can shift our numbering up one, with major releases for adding/removing Pythons, minor release for features, and patch releases for bug fixes. This would allow the much-requested `vX` GHA tag to be something we can provide.

Here, I'm less sure. Personally, I see the *compatibility* part of our versioning system to be related to the package and configuration, not so much the Python versions that we build. That's my logic behind not being too worried about bumping major when dropping an old version. I see the specific versions of wheels produced more of an side-effect of the time that version of cibuildwheel was released, not part of the API contract.

Having said all that, we don't follow Semantic versioning anyway, so maybe the specific definition of compatibility is moot.

The bigger point is - do users need to care? When I see a package I use bump major, I'll assume there might be some work I need to do to update - read the release notes, maybe update the config files as a result. But I don't think that applies for Python addition/removal updates. The Python additions require no work (assuming you've already made the package compatible during the beta phase), and the removals are probably so old as to not be of interest (we're still building 3.6!).

Put another way - should we be requiring a major version bump to move an old version of PyPy into the `pypy-eol` enable group? I think it's better to reserve the major bump for circumstances where packagers need to pay attention, outside of the usual churn of interpreter versions.

**Comment by joerick** [comment]
An aside, but since we're talking about a new major version, a couple things we should consider-

* Changing the default `build-frontend` to `build` - the only reason we're still on `pip` is inertia, in my opinion. Of course, the `build` method of doing sdist->wheel builds will break some users, who might have to switch back to `pip`, but I think `build` is a better default these days.
* Adding delvewheel as a default option for `repair-wheel-command` on Windows. (cc @adang1345)

I think I'd be in favour of both, but curious of opinions.

**Comment by henryiii** [comment]
> the build method of doing sdist->wheel builds

That won't affect us at all, since we build with `--wheel`, which is a direct build, like pip's. If we made SDists, they'd collide.

Yes, more changes is fine, I was just listing what I could quickly think of.

It might be a good time to finally switch the docs to show the TOML option first, with the envvar as second tab.

As for moving to more SemVer-like versioning, it was mostly to get people to stop complaining about a lack of a `vX` tag. Not required if we'd rather say with the current versioning scheme.

**Comment by joerick** [comment]
> That won't affect us at all, since we build with `--wheel`, which is a direct build, like pip's. If we made SDists, they'd collide.

Oh, I didn't know that. That's good. So `build`'s wheel builds are 'in-tree', like pip's then?

> It might be a good time to finally switch the docs to show the TOML option first, with the envvar as second tab.

Good call.

> As for moving to more SemVer-like versioning, it was mostly to get people to stop complaining about a lack of a `vX` tag. Not required if we'd rather say with the current versioning scheme.

My opinion on this has shifted lately... I think we should just relent and make a `vX` tag anyway. I think that while it's probably not a good idea for serious, professional packages, there are plenty of hobbyists who just want wheel builds and don't particularly care about the details. That seems valid to me. The number of side-projects I have that get constant dependabot spam for CVEs I don't care about, the dependabot fatigue is real.

**Comment by adang1345** [comment]
Regarding adding a `delvewheel` default repair command, I'm not sure. I have noticed that current `cibuildwheel` users frequently use the `--add-path` option to tell `delvewheel` where to find DLL dependencies during the repair process, and the path used is different for each project. It may be better to require the user to manually set the repair command and just put an example command in the documentation such as `delvewheel repair -w {dest_dir} {wheel}`.

**Comment by henryiii** [comment]
> So build's wheel builds are 'in-tree', like pip's then?

`pipx build` is not in-tree, it builds a wheel from an SDist. But `pipx build --wheel` is, as there's no SDist built.

**Comment by mayeut** [comment]
I added #2052 to the list for 2.22 per https://github.com/pypa/cibuildwheel/pull/1912#issuecomment-2440003656 comment

**Comment by EwoutH** [comment]
2.0 might also be an oppertunity to bump the default manylinux version:
- https://github.com/pypa/cibuildwheel/pull/1988

**Comment by joerick** [comment]
Would it make sense to create a `v3` branch we can start merging these new features into?

**Comment by henryiii** [comment]
Aren't we just about ready to release 2.22? We just have #2048 and #2052 left. Then we can move forward with main being version 3, and make a v2 branch for back ports.

**Comment by henryiii** [comment]
IMO, 2.22 is ready.

**Comment by joerick** [comment]
I think it's time to get working on v3.0. Because the utils refactor is gonna be very merge-conflicty, I'll hold off on that until there are fewer PRs waiting to go in. I think a good one to start with is #1912.

**Comment by mattip** [comment]
Any plans to drop older versions of Python? I think 3.6 can be dropped, maybe even 3.7. That would save some CI processing.

**Comment by mattip** [comment]
Hmm, maybe I am wrong. The [urllib3](https://pypistats.org/packages/urllib3) download statistics claim around 2% for python3.6 and 17% (!!!) for 3.7.

**Comment by mayeut** [comment]
> Any plans to drop older versions of Python? I think 3.6 can be dropped, maybe even 3.7. That would save some CI processing.

It's planned in manylinux: https://github.com/pypa/manylinux/issues/1671
The planning takes into account CI to some extent as this will allow to add Python 3.14 Beta 1 with almost no impact on CI times  (replace 3.6/3.7 with 3.14/3.14t).
cibuildwheel could drop before but probably unlikely.

The download [statistics for manylinux wheels](https://mayeut.github.io/manylinux-timeline/) (Linux is probably where 3.6/3.7 are most used) shows 0.2% downloads for 3.6 and 4.5% for 3.7 over all supported packages at the moment. 11.5% of packages support Python 3.6 (either directly, as py3 or abi3 wheels) and 25% support Python 3.7.

**Comment by henryiii** [comment]
Quick thought: maybe we could drop 3.6 and 3.7 in 2.0; that timeframe would likely also allow #1538 to make it in for 3.0 too. We'd need to do a 2.22 patch release before then to get #2268 and the manylinux update out sooner.

**Comment by joerick** [comment]
I think it would make sense to drop 3.6 and 3.7 in this v3.0 release, in fact, it's a good opportunity to do so.

@henryiii I'm happy to make a v2.x branch and do a point/minor release, I'd probably do a minor release for a new PyPy version. Looking at the recent PRs, [things I'd like to backport are](https://github.com/pypa/cibuildwheel/pulls?q=is%3Apr+merged%3A%3E2024-11-23T01%3A44%3A09-0500+label%3Aneeds-backport):

- #2268 
- #2135
- #2269 
- #2117 
- ...and a dependency update 

Is there anything else you'd like to see?

**Comment by mayeut** [comment]
Would 3.0 be a good time to drop support for manylinux1, manylinux2010, manylinux_2_24 & musllinux_1_1 support ? I don't know why I didn't think of this earlier (maybe related to the late drop of python 3.6/3.7, maybe not). This would probably require to add a warning in 2.x ?

**Comment by henryiii** [comment]
Someone could still work around it by adding the explicit image name, I assume? We'd only be dropping support for the shortcut setting the pins?

**Comment by mayeut** [comment]
> Someone could still work around it by adding the explicit image name, I assume? We'd only be dropping support for the shortcut setting the pins?

Yes, the images aren't going to be removed (at least for now, there might be some clean-up in which version are available at some point though). That being said, manylinux1 images won't be usable after python 3.9 EOL for the before_all step as we use the oldest non EOL python there.

Extended support for RHEL / OL 6 has ended in 2024, support for musl 1.1 / Alpine stopped on 01 May 2022.

So all in all, they should remain usable but unsupported / untested.

**Comment by joerick** [comment]
I think dropping support for these images would be a good idea, given the base OSs are unsupported. I can see two scenarios for dropping official support,

1. we could put a deprecation notice in the next cibuildwheel release, and drop a few weeks later in v3.0
2. we could deprecate in v3.0 and throw a warning on each build that these images will stop updating soon and strongly recommend that the user should update

I'm more inclined to (2). (1) has a pretty short deprecation period, for an option that I suspect is still in use here and there[^1]. (2) might be a bit friendlier to cibuildwheel users, while still giving manylinux the freedom to sunset them without worrying that many cibuildwheel users will be left behind.

It does somewhat depend on what the maintenance implications are for manylinux.

[^1]: do we have download/usage stats for the oci images somewhere? it would be useful to have some data on this...

**Comment by mayeut** [comment]
There are no maintenance implications for manylinux whatsoever as those images are already EOL. (the clean-up mentioned is unrelated to any decision here & won't affect cibuildwheel).

On cibuildwheel side, it's mostly CI time (less tests)  & documentation updates.

There are some download stats on quay.io, roughly, only speaking about x86_64, manylinux2014 is around 10-15k downloads/day, then manylinux_2_28 around 6k/d, musllinux_1_2 4k/d, manylinux_2_24 500/d, musllinux_1_1 500/d, manylinux1 <500/d, manylinux2010 250/d.

For python<3.8, the pip version was important w.r.t the manylinux standard. For >= 3.8, almost all default shipped pip support manylinux2014 (I'm seeing 0.1% for manylinux1 downloads).

Given v3.0 is a clear breaking change, I'd rather see (1) but given the short time frame, I'd understand going with (2).

**Comment by henryiii** [comment]
Regardless, we can go ahead and add a warning for the next patch release, and then decide if we should make it an error or wait longer.

**Comment by joerick** [comment]
> There are some download stats on quay.io, roughly, only speaking about x86_64, manylinux2014 is around 10-15k downloads/day, then manylinux_2_28 around 6k/d, musllinux_1_2 4k/d, manylinux_2_24 500/d, musllinux_1_1 500/d, manylinux1 <500/d, manylinux2010 250/d.

Those numbers for manylinux2010 and manylinux1 are smaller than I expected. That does alter my intuition a bit. I'd be okay fully dropping support if others are in favour.

**Comment by henryiii** [comment]
I'm okay to drop in 3.0. It's very easy to fix, and people expect a little churn when updating to a major version, especially if they are using an unmaintained image.

**Comment by di** [comment]
> There are some download stats on quay.io

Would it be helpful to see stats on wheel uploads for Python 3.6/3.7? The [public dataset](https://docs.pypi.org/api/bigquery/) should have enough information to do this, I'd be happy to put it together if you'd like.

**Comment by joerick** [comment]
Hey @di! I tend to look at mayeut's [manylinux-timeline](https://mayeut.github.io/manylinux-timeline/) for that kind of statistics. It's only manylinux, but that's not a bad constraint since nearly every packager is gonna release manylinux wheels, and it's good that it only considers platform wheels, since that's the segment that cibuildwheel cares about.

**Comment by joerick** [comment]
While we're looking at manylinux and v3.0, I understand that manylinux2014 is on Centos 7 which is [EOL now](https://endoflife.date/centos) as well. Would it be worth bumping the default?

If so would `manylinux_2_28` make sense, given the issues I read about 2_34's tendency to produce  binaries incompatible with some x86_64 machines?

**Comment by joerick** [comment]
I confess I forgot about the discussion in #1772 - we had planned (and communicated) to drop it on 2025-06-30 due to an AmazonLinux EOL, but that EOL date has since changed to 2026. Maybe v3.0 is a better time to make the switch anyway? It would only be a couple months sooner than planned.

**Comment by henryiii** [comment]
I think 3.0 would be a nice time to do it. Will defer to @mayeut, though.

What about dropping seeding setuptools and wheel on Python 3.8-3.11? I think it would be better if users specify every package they actually need (ideally via pyproject.toml), and they already have to do this to support 3.12+. If we did that, we could drop seeding anything into the venvs, which would be simpler and faster.

**Comment by mayeut** [comment]
> I confess I forgot about the discussion in https://github.com/pypa/cibuildwheel/issues/1772 - we had planned (and communicated) to drop it on 2025-06-30 due to an AmazonLinux EOL, but that EOL date has since changed to 2026. Maybe v3.0 is a better time to make the switch anyway? It would only be a couple months sooner than planned.

it was well communicated in advance. Even though the EOL date changed, I'm not sure we should change the plan to change the default and 3.0 seems like a good time to do this. Usage is continuously dropping, especially if we consider python 3.7 being dropped in 3.0. It's hard to guess the real impact downstream, but the switch is likely the good thing to do here.

> What about dropping seeding setuptools and wheel on Python 3.8-3.11? I think it would be better if users specify every package they actually need (ideally via pyproject.toml), and they already have to do this to support 3.12+. If we did that, we could drop seeding anything into the venvs, which would be simpler and faster.

agree

**Comment by ngoldbaum** [comment]
@henryiii PyCon sprints start tomorrow. No pressure at all if you have other priorities, but doing a release might unblock people working on uploading nightly wheels. I'll be working on NumPy nightlies ASAP.

**Comment by henryiii** [comment]
We have a few things left that are almost done, maybe we could do a beta release though?

**Comment by henryiii** [comment]
I might actually stop by the sprints in the morning.

**Comment by joerick** [comment]
@ngoldbaum I just released [v3.0.0b1](https://github.com/pypa/cibuildwheel/releases/tag/v3.0.0b1)! Hope the sprints go well!

**Comment by henryiii** [comment]
If it's helpful, I'm sticking the changelog here so I can merge all the parts, and we could also edit it if a maintainer wants to (caveat being that @joerick might prefer to put it together another way and skip what's here ;) ):

This does _not_ include anything after the current changelog (rc3). I've updated the option syntax to the new standard. Added mention of verbosity reworking. A bunch of PR numbers fixed due to GitHub showing previews. :)

---


#### v3.0.0

_date_

See the [release post](https://iscinumpy.dev/post/cibuildwheel-3-0-0/) for more info on new features!

- 🌟 Adds the ability to [build wheels for iOS](https://cibuildwheel.pypa.io/en/latest/platforms/#ios)! Set the [`platform` option](https://cibuildwheel.pypa.io/en/latest/options/#platform) to `ios` on a Mac with the iOS toolchain to try it out! (#2286, #2363, #2432)
- 🌟 Adds support for the GraalPy interpreter! Enable for your project using the [`enable` option](https://cibuildwheel.pypa.io/en/latest/options/#enable). (#1538, #2411, #2414)
- ✨ Adds CPython 3.14 support, under the [`enable` option](https://cibuildwheel.pypa.io/en/latest/options/#enable) `cpython-prerelease`. This version of cibuildwheel uses 3.14.0b2. (#2390)

    _While CPython is in beta, the ABI can change, so your wheels might not be compatible with the final release. For this reason, we don't recommend distributing wheels until RC1, at which point 3.14 will be available in cibuildwheel without the flag._ (#2390)
- ✨ Adds the [test-sources option](https://cibuildwheel.pypa.io/en/latest/options/#test-sources), and changes the working directory for tests. (#2062, #2284, #2437)

    - If this option is set, cibuildwheel will copy the files and folders specified in `test-sources` into the temporary directory we run from. This is required for iOS builds, but also useful for other platforms, as it allows you to avoid placeholders.
    - If this option is not set, behaviour matches v2.x - cibuildwheel will run the tests from a temporary directory, and you can use the `{project}` placeholder in the `test-command` to refer to the project directory. (#2420)

- ✨ Adds `dependency-versions` inline syntax (#2122)
- ✨ Improves support for Pyodide builds and adds the experimental [`pyodide-version`](https://cibuildwheel.pypa.io/en/latest/options/#pyodide-version) option, which allows you to specify the version of Pyodide to use for builds. (#2002)
- ✨ Add `pyodide-prerelease` enable option, with an early build of 0.28 (Python 3.13). (#2431)
- ✨ Adds the [`test-environment`](https://cibuildwheel.pypa.io/en/latest/options/#test-environment) option, which allows you to set environment variables for the test command. (#2388)
- ✨ Adds the [`xbuild-tools`](https://cibuildwheel.pypa.io/en/latest/options/#xbuild-toolss) option, which allows you to specify tools safe for cross-compilation. Currently only used on iOS; will be useful for Android in the future. (#2317)
- 🛠 The default [manylinux image](https://cibuildwheel.pypa.io/en/latest/options/#linux-image) has changed from `manylinux2014` to `manylinux_2_28`. (#2330)
- 🛠 EOL images `manylinux1`, `manylinux2010`, `manylinux_2_24` and `musllinux_1_1` can no longer be specified by their shortname. The full OCI name can still be used for these images, if you wish. (#2316)
- 🛠 32-bit linux builds removed from `"auto"`, requires explicit `"auto32"`. Note that modern manylinux images (like `manylinux_2_28`) do not have 32-bit versions. (#2458)
- 🛠 Invokes `build` rather than `pip wheel` to build wheels by default. You can control this via the [`build-frontend`](https://cibuildwheel.pypa.io/en/latest/options/#build-frontend) option. You might notice that you can see your build log output now! (#2321)
- 🛠 Build verbosity settings have been reworked to have consistent meanings between build backends when non-zero. (#2339)
- 🛠 Removed the `CIBW_PRERELEASE_PYTHONS` and `CIBW_FREE_THREADED_SUPPORT` options - these have been folded into the [`enable`](https://cibuildwheel.pypa.io/en/latest/options/#enable) option instead. (#2095)
- 🛠 Build environments no longer have setuptools and wheel preinstalled. (#2329)
- 🛠 Use the standard Schema line for the integrated JSONSchema. (#2433)
- ⚠️ Dropped support for building Python 3.6 and 3.7 wheels. If you need to build wheels for these versions, use cibuildwheel v2.23.3 or earlier. (#2282)
- ⚠️ The minimum Python version required to run cibuildwheel is now Python 3.11. You can still build wheels for Python 3.8 and newer. (#1912)
- ⚠️ PyPy wheels no longer built by default, due to a change to our options system. To continue building PyPy wheels, you'll now need to set the [`enable` option](https://cibuildwheel.pypa.io/en/latest/options/#enable) to `pypy` or `pypy-eol`. (#2095)
- ⚠️ Dropped official support for Appveyor. If it was working for you before, it will probably continue to do so, but we can't be sure, because our CI doesn't run there anymore. (#2386)
- 📚 Use Python 3.14 color output in docs CLI output. (#2407)
- 📚 A reorganisation of the docs, and numerous updates. (#2280)
- 📚 Docs now primarily use the pyproject.toml name of options, rather than the environment variable name. (#2389)
- 📚 README table now matches docs and auto-updates. (#2427, #2428)

**Comment by hugovk** [comment]
Thank you for including 3.14!
  
> * For this reason, we don't recommend distributing wheels until RC1, at which point 3.14 will be available in cibuildwheel without the flag.

In the [CPython release notes](https://discuss.python.org/t/python-3-14-0-beta-2-is-here/93396), we now encourage creating pre-releases for the beta:

> This includes creating pre-release wheels for 3.14, as it helps other projects to do their own testing. However, we recommend that your regular production releases wait until 3.14.0rc1, to avoid the risk of ABI breaks.

**Comment by henryiii** [comment]
The old recommendation was beta 4, is "ABI stable" moved to RC 1 now? Practically, I think most people wait till RC 1 anyway, and that's when cibuildwheel enables it by default, but I remember that beta 4 was the official stability starting point.

Most projects don't make prereleases, and when they do, it's based on the project, not on CPython's release schedule. And if you added it to prereleases, you'd have to pull it back out of real releases. And with so many binary platforms these days (CPython's doubled the number of binaries required since 3.13), projects shipping binaries can't make too many releases before running out of PyPI space. So it's not a bad recommendation, but I don't see it being something most projects can follow.

There is a solution for the Scientific Python projects: [SPEC 4: Nightly wheels](https://scientific-python.org/specs/spec-0004/). By adding https://pypi.anaconda.org/scientific-python-nightly-wheels/simple as an extra index to pip/uv and enabling pre-releases, you can get nightlies from a variety of projects. These nightly development releases include Python 3.14 as soon as upstream supports it (well, by that night, anyway); [numpy](https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/numpy/) shipped 3.14 (and pypy3.11) wheels there, we got them in during the PyCon sprints. I'm going to be adding documentation about this to cibuildwheel soon, as it's a great way to start testing projects that depend on these (and projects can ask for an account there). It might be nice if there was an official "nightlies" PyPI. (The key feature is that old releases are deleted, I think it's "last three releases" or something like that).

**Comment by joerick** [comment]
> If it's helpful, I'm sticking the changelog here so I can merge all the parts, and we could also edit it if a maintainer wants to (caveat being that [@joerick](https://github.com/joerick) might prefer to put it together another way and skip what's here ;) ):

Looks great, thank you! :)

**Comment by joerick** [comment]
I think we're ready for a (final!) rc.

**Comment by henryiii** [comment]
Updated the above changelog for 3.0.0rc3, and released RC 3! Unless something else pops up, this likely will be the release version (minus docs/ci tweaks).

* https://github.com/scikit-hep/boost-histogram/pull/1016

**Comment by joerick** [comment]
Aaaaand it's out! 🎉 

https://github.com/pypa/cibuildwheel/releases/tag/v3.0.0

Congrats all for getting this over the line!
